src/back/glsl_common.rs

```diff
@@ -0,0 +1,704 @@
+                                width
+                            ))),
+                    },
+                    ScalarKind::Bool => "b",
+                },
+                *columns as u8,
+                *rows as u8
+            ),
+            TypeInner::Pointer { base, class } => todo!(),
+            TypeInner::Array { base, size, stride } => format!(
+                "{}[{}]",
+                base.write_glsl(types, structs)?,
+                size.write_glsl()
+            ),
+            TypeInner::Struct { .. } => structs.get(self).unwrap().0.clone(),
+            TypeInner::Image { base, dim, flags } => format!(
+                "{}texture{}{}",
+                match types[*base].inner {
+                    TypeInner::Scalar { kind, .. } => match kind {
+                        ScalarKind::Sint => "i",
+                        ScalarKind::Uint => "u",
+                        ScalarKind::Float => "",
+                        _ => return Err(Error::Custom(format!("Cannot build image of booleans",))),
+                    },
+                    _ =>
+                        return Err(Error::Custom(format!(
+                            "Cannot build image of type {}",
+                            base.write_glsl(types, structs)?
+                        ))),
+                },
+                dim.write_glsl(),
+                flags.write_glsl()
+            ),
+            TypeInner::DepthImage { dim, arrayed } => format!(
+                "texture{}{}",
+                dim.write_glsl(),
+                if *arrayed { "Array" } else { "" }
+            ),
+            TypeInner::Sampler { comparison } => String::from(if *comparison {
+                "sampler"
+            } else {
+                "samplerShadow"
+            }),
+        })
+    }
+}
+
+impl StorageClass {
+    pub(crate) fn write_glsl<'a>(&'a self) -> impl fmt::Display + 'a {
+        struct StorageClassWriter<'a> {
+            inner: &'a StorageClass,
+        }
+
+        impl<'a> fmt::Display for StorageClassWriter<'a> {
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                write!(
+                    f,
+                    "{}",
+                    match self.inner {
+                        StorageClass::Constant => "const ",
+                        StorageClass::Function => todo!(),
+                        StorageClass::Input => "in ",
+                        StorageClass::Output => "out ",
+                        StorageClass::Private => "",
+                        StorageClass::StorageBuffer => "buffer ",
+                        StorageClass::Uniform => "uniform ",
+                        StorageClass::WorkGroup => "shared ",
+                    }
+                )
+            }
+        }
+
+        StorageClassWriter { inner: self }
+    }
+}
+
+impl ArraySize {
+    pub(crate) fn write_glsl<'a>(&'a self) -> impl fmt::Display + 'a {
```

Comment:
I see the pattern now. I suggest we do the following instead:
```rust
struct ArraySize(crate::ArraySize);
impl fmt::Disaplay for ArraySize {..}
```
I believe it's also how WGSL backend does this.